### PR TITLE
utilize `@branchHint` more

### PR DESCRIPTION
**Pull request by Rexicon226**

### CODE REVIEW

File: lib/std/hash_map.zig [hunk 1]
```diff
@@ -1187,17 +1187,13 @@ pub fn HashMapUnmanaged(
...
-        /// fuse the basic blocks after the branch to the basic blocks
-        /// from this function.  To encourage that, this function is
-        /// marked as inline.
-        inline fn getIndex(self: Self, key: anytype, ctx: anytype) ?usize {
+        fn getIndex(self: Self, key: anytype, ctx: anytype) ?usize {
             comptime verifyContext(@TypeOf(ctx), @TypeOf(key), K, Hash, false);
 
             if (self.size == 0) {
+                // We use cold instead of unlikely because even with unlikely,
+                // LLVM will want to fallthrough to this case.
```
Comment (andrewrk) [inline lib/std/hash_map.zig:1195]: This is not LLVM-specific code so this justification should not mention LLVM.

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: lib/std/hash_map.zig (hunk 1 above)
Comment (Rexicon226) [reply lib/std/hash_map.zig]: Good point, I've updated it to explain the specific behavior I'm looking for.

Comment (andrewrk) [reply lib/std/hash_map.zig]: Looks good, thanks!

**Comment by mlugg** [comment]
Note that the LLVM backend doesn't currently do anything with function-level branch hints which aren't `.cold`. @Rexicon226, if you know of a sane way we could lower `likely`/`unlikely` without doing the work of propagating the hint to callsites ourselves, please do open an issue. (Otherwise, this will hopefully be improved once we implement our own optimization passes.)

**Comment by rohlem** [comment]
> if you know of a sane way we could lower `likely`/`unlikely` without doing the work of propagating the hint to callsites ourselves

Uninformed drive-by idea, does `if (true)` count as a branch? Could we wrap the entire function body with `if (true) { ... }` and place the branch hint on that?
(Optimizations might still only happen when LLVM inlines at the callsite, but I think that's kind of expected - if functions are analyzed in isolation, and we don't manually propagate the branch hint information, then there's no information to act on.)

**Comment by mlugg** [comment]
All that would do is tell LLVM that the *body* of the `if (true)` was likely to be reached; which is obvious, because it's guaranteed! In other words, such a hint would be a nop.

**Comment by Rexicon226** [comment]
> If you know of a sane way we could lower likely/unlikely without doing the work of propagating the hint to callsites ourselves, please do open an issue.

Hmm, I did not consider that. I'll need to think about it. Afaik, no LLVM function attribute other than `cold` affects this sort of thing, and `cold` does something different than what `unlikely` does. I'll leave the code as is since this is LLVM's fault and not the code's :)

> Optimizations might still only happen when LLVM inlines at the callsite, but I think that's kind of expected

To add to what mlugg said, this isn't the case. One of the key points of the `cold` attribute is that it prevents LLVM from inlining the function as well as influencing branch weight. The same thing should apply to an `unlikely` function, doing 
```zig
if (x) {
    unlikely();
} else {
    likely();
}
```
Should clearly fallthrough to the `likely` function call.

**Pull request merged**